### ISSUE TITLE
Autoomatically set dropout for SDPA depending on training mode

### ISSUE THREAD

**Issue by fxmarty** [pytorch/pytorch#99293]
### 🚀 The feature, motivation and pitch

Hi,

`torch.nn.functional.dropout` has a `training` argument that allows to disable/enable dropout during inference. In a module, it simplifies code by passing simply `training=self.training` instead of `p=self.droupout_p if self.training else 0.0`. This avoids as well a python controlflow (that dynamo may not like).

```python
a = torch.rand(20, 30)
b = torch.nn.functional.dropout(a, p=0.5, training=False)
```

It would be great to have a similar option in `torch.nn.scaled_dot_product_attention`, as e.g. `eval()` does not change the behavior (dummy example):

```python
import torch
import torch.nn as nn

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()

        self.q = nn.Parameter(torch.rand(1, 64, 32, 64, dtype=torch.float16).to("cuda"))
        self.k = nn.Parameter(torch.rand(1, 64, 32, 64, dtype=torch.float16).to("cuda"))
        self.v = nn.Parameter(torch.rand(1, 64, 32, 64, dtype=torch.float16).to("cuda"))
        
    def forward(self, x):
        res = torch.nn.functional.scaled_dot_product_attention(self.q, self.k, self.v, is_causal=True, dropout_p=0.8) * x
        return res

model = MyModel()
model = model.eval()
inp = torch.Tensor([3.]).to("cuda", torch.float16)
with torch.inference_mode():
    res = model(inp)

print(res)
"""
prints
tensor([[[[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          ...,
          [1.0000, 0.4944, 0.9648,  ..., 0.9160, 0.9844, 1.2383],
          [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [3.4277, 1.6973, 2.8184,  ..., 3.3672, 2.6348, 2.9453]],
...
"""
```

cc @drisspg @cpuhrsch @mikekgfb

### Alternatives

Adding a controlflow

### Additional context

/
